### ISSUE TITLE
Account for relations implied by title

### ISSUE THREAD

**Issue by fkohrt** [flibbles/tw5-relink#25]
I had this idea that people who express relationships between tiddlers through their titles might also benefit from this plugin. Say someone organizes their tiddlers in a hierarchical way mimicking their file system:

```
    Images
    ├── ...
    ├── Images/2019
    │   ├── Images/2019/01-Trip-to-Spain
    │   ├── Images/2019/05-School
    │   └── Images/2019/12-Christmas
    └── ...
```

They might use the [explorer](https://tiddlywiki.com/#%24%3A%2Fcore%2Fui%2FMoreSideBar%2FExplorer)'s [tree macro](https://tiddlywiki.com/#tree%20Macro) to discover their tiddlers.

Then renaming the tiddler `Images` surely should also rename `Images/2019`, `Images/2019/01-Trip-to-Spain`, `Images/2019/05-School`, `Images/2019/12-Christmas` and so on. [TiddlyResearch](https://kebifurai.github.io/TiddlyResearch/) actually has the notion of "folders" but internally they are implemented via tags.

I can imagine other conventions such as storing other formats of tiddlers right "beneath" them

```
    History
    ├── History/LaTeX
    └── History/PDF
```

or translating tiddlers with the [MediaWiki convention](https://www.mediawiki.org/wiki/Help:Extension:Translate/Page_translation_administration#Anatomy_of_a_translatable_page)

```
    History
    ├── History/de
    ├── History/fr
    └── History/it
```

This feature (renaming all tiddlers that are connected by their titles) may have to be an opt-in option, though.

**Comment by flibbles** [comment]
So I actually started creating a subplugin for doing something like this. In their case, they wanted to rename system tiddlers that referenced the renamed tiddler. So renaming `myTiddler` would also rename `$:/config/Visibility/myTiddler`. That issue petered out, but I still have the framework of what I did. My stopping point had been how to do so that makes it useful in all *derived title* cases.

So you get your: Changing `History` renames `History/LaTeX`. but presumably not: Changing `History` renames `History_channel`

But somebody else *might* want renames to happen using '_' as a delimiter because they're loading a NodeJS tiddlywiki that loads all the files from a flat directory.

But it'd also still allow the person who wants to rename configuration tiddlers like I described above.

My first thought was to have some sort of configuration filter field you could fill out.  
Like you could put in `[removeprefix<fromTiddler>prefix[/]addprefix<toTiddler>]`.  
Where as the underscore guy could put in `[removeprefix<fromTiddler>prefix[_]addprefix<toTiddler>]`.  
And then the configuration guy could have `[is[system]removesuffix<fromTiddler>addsuffix<toTiddler>]`.  

Basically, it would work like FileSystemPath. Where Relink would input each tiddler into the filter you created, and if something comes out the other side, that result is used as a new name. But I was concerned that this was over elaborate. Does this seem like the kind of thing that could work for you?

**Comment by fkohrt** [comment]
Oh that's cool!

I'm not sure if I already understood how the usage would look like. So in a use case where I would like to rename the `History` tiddler, I would first have to fill the configuration filter field with one of the expressions you provided. Are the variables `fromTiddler` and `toTiddler` set automatically by the plugin? Then I rename the `History` tiddler. Do I have to clear the configuration filter field before I use Relink again (such as, for updating a tag tiddler)? And how would I solve cases where the renamed tiddler is neither at the top nor at the bottom, such as in `Subjects/History/de`?

**Comment by flibbles** [comment]
Okeydokey!!!

Now that three months have past, and I'm sure you've completely forgotten about this, I've created the Relink-titles supplemental plugin, which I believe satisfies your problem (and a few other people's problems too).

It's not officially released yet. I think I may still make some changes. Particularly, there's no way for 3rd party plugins to register title patterns for themselves and have them automatically take effect. That's because my current implementation boils all the title renaming down to a single filter which plugins would be able to modify easily.

Check it out [here (The "Titles" tab)](https://flibbles.github.io/tw5-relink/#Plugins) and let me know what you think.

**Comment by arunnbabu81** [comment]
I think this supplemental plug in is useful for those people who uses streams plug ins. This helps in automatically renaming the child nodes of the parent streams tiddler if we make any change in the title of the parent streams tiddler. Thanks for this add on. When can we expect this to release officially?

**Comment by flibbles** [comment]
Probably a few days. Right now, it works by conglomerating all your rename rules into a single filter, which I don't think works. I mean, it _works_, but take the streams plugin. Ideally, the author would want some way for his plugin and Relink-titles to integrate automatically, but there's no declarative way to do that right now.

I think I need to make Relink-titles work by keeping distinct filters separate. That way, there isn't some conglomerate filter that needs to be manually assembled.

Thank you for pointing out streams, by the way. It seems they're already going to some effort to integrate with Relink, which makes it all the more important to me that Relink-titles integration is possible too.

**Comment by fkohrt** [comment]
> Check it out [here (The "Titles" tab)](https://flibbles.github.io/tw5-relink/#Plugins) and let me know what you think.

Awesome, just what I imagined – I really like it. Thank you very much for making it possible! Feel free to close this issue once you consider it ready :)

**Comment by flibbles** [comment]
Okay. _Relink-titles_ is released. It currently only has one rule, and that is to update nested tiddlers in a filesystem-like wiki, but other rules are easily added. ...As soon as I think of any. But also third party plugins can easily add their own rules too.

Make sure you use the official release, and not the test one I released a few days ago. It had some bad bugs that could delete swaths of tiddlers under some circumstances. I fixed them all, but you know... it's always a good idea anyway to back up your project when installing a new plugin.

**Comment by flibbles** [comment]
Also, like it says all over the plugin, you __must__ update Relink to 1.10.2 to use this, or those bugs I was describing can come up. This was unavoidable. Relink was never prepared to rename titles it was currently relinking. The core code needed some small but important changes.
